snapshot/archive.go

```diff
@@ -193,8 +194,11 @@ func read(in io.Reader, metadata *raft.SnapshotMeta, snap io.Writer) error {
 
 		switch hdr.Name {
 		case "meta.json":
-			dec := json.NewDecoder(io.TeeReader(archive, metaHash))
-			if err := dec.Decode(&metadata); err != nil {
+			buf, err := ioutil.ReadAll(io.TeeReader(archive, metaHash))
```

Comment:
I think it's worth commenting on why this is needed. The root cause seems to be something like:

```
// json.Decode will stop reading as soon as the initial JSON element (the object) is closed 
// (i.e. at the closing brace) which means the final newline, if any, is not read and so is not 
// delivered to `metaHash`. This means that the hash calculated is not the same as the
// one in SHA256SUMS which was calculated over the raw bytes in the file. By explicitly 
// reading the whole thing first we ensure that we calculate the hash over the exact
// same bytes regardless of whitespace.
```

Reply:
Good point, I will add a comment!